src/mongocxx/lib/mongocxx/v1/uri.cpp

```diff
@@ -12,4 +12,411 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include <mongocxx/v1/uri.hpp>
+#include <mongocxx/v1/uri.hh>
+
+//
+
+#include <bsoncxx/v1/document/view.hpp>
+#include <bsoncxx/v1/stdx/optional.hpp>
+#include <bsoncxx/v1/stdx/string_view.hpp>
+#include <bsoncxx/v1/types/view.hpp>
+
+#include <mongocxx/v1/exception.hh>
+#include <mongocxx/v1/read_concern.hh>
+#include <mongocxx/v1/read_preference.hh>
+#include <mongocxx/v1/write_concern.hh>
+
+#include <cstdint>
+#include <string>
+#include <system_error>
+#include <vector>
+
+#include <bsoncxx/private/bson.hh>
+#include <bsoncxx/private/immortal.hh>
+
+#include <mongocxx/private/mongoc.hh>
+#include <mongocxx/private/scoped_bson.hh>
+#include <mongocxx/private/utility.hh>
+
+#include <bson/error.h>
+
+namespace mongocxx {
+namespace v1 {
+
+using code = uri::errc;
+
+namespace {
+
+mongoc_uri_t* to_mongoc(void* ptr) {
+    return static_cast<mongoc_uri_t*>(ptr);
+}
+
+template <typename T>
+bsoncxx::v1::stdx::optional<T> get_field(mongoc_uri_t const* uri, char const* field);
+
+template <>
+bsoncxx::v1::stdx::optional<std::int32_t> get_field(mongoc_uri_t const* uri, char const* field) {
+    auto const options = libmongoc::uri_get_options(uri); // Never null.
+
+    bson_iter_t iter = {};
+    if (bson_iter_init_find_case(&iter, options, field) && bson_iter_type(&iter) == BSON_TYPE_INT32) {
+        return bson_iter_int32(&iter);
+    }
+
+    return {};
+}
+
+template <>
+bsoncxx::v1::stdx::optional<bool> get_field(mongoc_uri_t const* uri, char const* field) {
+    auto const options = libmongoc::uri_get_options(uri); // Never null.
+
+    bson_iter_t iter = {};
+    if (bson_iter_init_find_case(&iter, options, field) && bson_iter_type(&iter) == BSON_TYPE_BOOL) {
+        return bson_iter_bool(&iter);
+    }
+
+    return {};
+}
+
+template <>
+bsoncxx::v1::stdx::optional<bsoncxx::v1::stdx::string_view> get_field(mongoc_uri_t const* uri, char const* field) {
+    auto const options = libmongoc::uri_get_options(uri); // Never null.
+
+    bson_iter_t iter = {};
+    if (bson_iter_init_find_case(&iter, options, field) && bson_iter_type(&iter) == BSON_TYPE_UTF8) {
+        std::uint32_t len = {};
+        auto const data = bson_iter_utf8(&iter, &len);
+        return bsoncxx::v1::stdx::string_view{data, len};
+    }
+
+    return {};
+}
+
+template <typename T>
+bsoncxx::v1::stdx::optional<T> get_credentials_field(mongoc_uri_t const* uri, char const* field);
+
+template <>
+bsoncxx::v1::stdx::optional<bsoncxx::v1::document::view> get_credentials_field(
```

Comment:
Remove unnecessary template? I expect it is not needed since there is only one specialization.

```suggestion
bsoncxx::v1::stdx::optional<bsoncxx::v1::document::view> get_credentials_field(
```